_source/_posts/2021-10-08-private-key-jwt-python.md

```diff
@@ -0,0 +1,165 @@
+
+But let's break it down a bit more:
+
+### Step 1
+
+Typically, one may think of JWTs as strictly being in the realm of the Identity Provider (IdP) - and in most cases that's correct. When an IdP grants tokens to a service provider (SP), it is signing those tokens with a private/secret key. Part of the responsibility of the SP when validating tokens is to use a public key provided by the IdP to verify the signature.
+
+The Private Key JWT flow flips this on its head - at least at the start. It's now the SP's turn to generate a public/private keypair, share the public key with the IdP (more on that in Step 2), create a JWT and sign it with their private key (more on that in Step 3).
+
+But I digress...Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048 bit key size).
+
+### Step 2
+
+The next step is registering your public key with the IdP - with Okta, that's done through our [clients](https://developer.okta.com/docs/reference/api/oauth-clients/#register-new-client) API. With this step, Okta will create an application integration and generate a unique client ID - this value will be important when it comes time to create the Private Key JWT. It will be this client ID as well as the working public key decryption that validates your request.
+
+In other flows like the popular [Authorization Code](https://developer.okta.com/docs/guides/implement-grant-type/authcode/main) flow, the identity provider generates a secret value called the client secret. In the Private Key JWT flow, the burden is on the service provider to generate a secret value for the same purpose - albeit by different means.
+
+### Step 3
+
+Technically this could be broken down into two separate steps - creating the private key JWT and then submitting it to the IdP to obtain an access token. I lumped it into one step because the Python command `get_access_token.py` below does both of these at once.
+
+Using the private key created in step 1, you sign a JWT with the following claims:
+
+```
+'iss': {client ID generated in step 2},
+'sub': {client ID generated in step 2},
+'aud': {okta_url}/oauth2/v1/token,
+"exp": {epoch expiration at future time}
+```
+
+This token is sent to the /token endpoint of the IdP:
+
+```
+  POST /token HTTP/1.1
+  Host: server.example.com
+  Content-Type: application/x-www-form-urlencoded
+
+  grant_type=client_credentials&
+    scope={scopes}&
+    client_id={client ID generated in step 2}&
+    client_assertion_type=
+    urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&
+    client_assertion={private key JWT}
+```
+
+Assuming everyting is setup correctly, the IdP should respond with an access token that you can use to access scoped resources.
+
+For an even deeper look at this flow, you can check out our walkthrough [here](https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/overview/). For now, it's time to set this up with Python!
```

Comment:
```suggestion
For an even deeper look at this flow, you can [check out our walkthrough](https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/overview/). For now, it's time to set this up with Python!
```
Style guide best practice: We avoid use of "here" "click this" etc - instead use a word or short phrase that describes more or less where the link will go.